### PR TITLE
Make error message consistent when invalid script is passed to -File, better error when passed ambiguous arg

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
Since -File is now the positional parameter for powershell.exe, made the error message consistent with 
-Command when passed an invalid file.

If ambiguous arg is passed, give a better error message:

```powershell
PS C:\users\poker\repos\PowerShell> powershell -no -file foo
Invalid argument '-no', did you mean:

        -nologo
        -noexit
        -noprofile
        -noninteractive
PowerShell v6.0.0-beta.5-42-g3ed9a816b2452551aaad1ac1245a1cbb35e87612-dirty
Copyright (C) Microsoft Corporation. All rights reserved.
```

Enable -WindowStyle to work

Fix https://github.com/PowerShell/PowerShell/issues/4351

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx [hunk 1]
```diff
@@ -252,7 +252,7 @@ EXAMPLES
     <value>Processing -File '{0}' failed because the file does not have a '.ps1' extension. Specify a valid Windows PowerShell script file name, and then try again.</value>
   </data>
   <data name="ArgumentFileDoesNotExist" xml:space="preserve">
-    <value>The argument '{0}' to the -File parameter does not exist. Provide the path to an existing '.ps1' file as an argument to the -File parameter.</value>
+    <value>The term '{0}' is not recognized as the name of a script file. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.</value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx]: I believe `argument` is better `term` and we can be shorter:
```
The argument '{0}' is not recognized as the name of a script file.
```

Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx]: Will change

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 2]
```diff
@@ -930,6 +954,27 @@ bool TryGetBoolValue(string arg, out bool boolValue)
 
                 if (!System.IO.File.Exists(_file))
                 {
+                    if (args[i].StartsWith("-") && args[i].Length > 1)
+                    {
+                        string param = args[i].Substring(1, args[i].Length - 1);
+                        StringBuilder possibleParameters = new StringBuilder();
+                        foreach (string validParameter in validParameters)
+                        {
+                            if (validParameter.Contains(param))
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Should this be `validParameter.StartsWith(param, StringComparer.OrdinalIgnoreCase)`?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 3]
```diff
@@ -930,6 +954,27 @@ bool TryGetBoolValue(string arg, out bool boolValue)
...
+                            {
+                                possibleParameters.Append("\n\t-");
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I'd indent with spaces - we don't indent with tabs anywhere else.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 4]
```diff
@@ -930,6 +954,27 @@ bool TryGetBoolValue(string arg, out bool boolValue)
...
+                                possibleParameters.Append(validParameter);
+                            }
+                        }
+                        if (possibleParameters.Length > 0)
+                        {
+                            WriteCommandLineError(
+                                string.Format(CultureInfo.CurrentCulture, CommandLineParameterParserStrings.InvalidArgument, args[i]),
+                                showBanner: false);
+                            WriteCommandLineError(possibleParameters.ToString(), showBanner: true);
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Showing the banner after the error is weird and not like the other errors I reviewed.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 2 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I was thinking to cover cases like: `-format`

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 3 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: will change

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 2 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I'll add a test case for this

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 4 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I thought that was weird in general, but it seems consistent with other uses of `WriteCommandLineError`.  I'm assuming it's to show that powershell still started if you started it within powershell since the prompt would look same/similar.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 5]
```diff
@@ -169,6 +169,30 @@ public override void WriteWarningLine(string message)
...
     {
+        internal static string[] validParameters = {
+            "psconsoleFile",
+            "version",
+            "nologo",
+            "noexit",
+#if !CORECLR
+            "sta",
+            "mta",
+#endif
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: `#if !CORECLR ` is no longer relevant. Should we remove the parameters at all?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 6]
```diff
@@ -446,7 +446,6 @@ internal enum KeyboardFlag : uint
         internal const int SW_MAX = 11;
 
 
-#if !CORECLR  // ProcessWindowStyle does Not exist on CoreCLR
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: Maybe 
`#if !UNIX`

File: src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx [hunk 7]
```diff
@@ -269,4 +269,7 @@ EXAMPLES
   <data name="MissingConfigurationNameArgument" xml:space="preserve">
     <value>Cannot process the command because -Configuration requires an argument that is a remote endpoint configuration name. Specify this argument and try again.</value>
   </data>
+  <data name="InvalidArgument" xml:space="preserve">
+    <value>Invalid argument '{0}', did you mean:</value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx]: "did you mean:" - Do you want add something else?

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 8]
```diff
@@ -492,6 +492,85 @@ foo
...
+                $outString | Should Match $expectedMatch
+            }
+        }
+    }
+}
+
+Describe "WindowStyle argument" -Tag Feature {
+    BeforeAll {
+        $ExitCodeBadCommandLineParameter = 0xFFFD0000
+        Add-Type -Name User32 -Namespace Test -MemberDefinition @"
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: We should exclude this on Unix.

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 9]
```diff
@@ -492,6 +492,85 @@ foo
...
+{
+    Hidden = 0,
+    Normal = 1,
+    Minimized = 2,
+    Maximized = 3,
+}
+"@
+    }
+
+    It "-WindowStyle <WindowStyle> should work on Windows" -Skip:(!$IsWindows) -TestCases @(
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: All tests is excluded on Unix so we could use our pattern "$PSDefaultParameterValues["it:skip"] = ..."

**Review by iSazonov — COMMENTED** [review]
Leave a comment

File: src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx (hunk 7 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx]: The possible matches show up after this.  See example above in the PR description.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 5 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I'm fixing that as part of enabling ApartmentState PR and not part of this PR

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs (hunk 6 above)
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: makes sense

File: test/powershell/Host/ConsoleHost.Tests.ps1 (hunk 8 above)
Comment (SteveL-MSFT) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: will do

File: test/powershell/Host/ConsoleHost.Tests.ps1 (hunk 9 above)
Comment (SteveL-MSFT) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: agree

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 5 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Closed.

File: src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx (hunk 7 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx]: Closed.

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 10]
```diff
@@ -492,6 +492,94 @@ foo
...
+                $outString | Should Match $expectedMatch
+            }
+        }
+    }
+}
+
+Describe "WindowStyle argument" -Tag Feature {
+    BeforeAll {
+        $defaultParamValues = $PSDefaultParameterValues.Clone()
+        $PSDefaultParameterValues["it:skip"] = (!$IsWindows)
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: Parentheses can be removed.

Comment (SteveL-MSFT) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: Will remove

**Comment by SteveL-MSFT** [comment]
Not sure why some of the new tests fail in AppVeyor but not on my desktop

**Comment by iSazonov** [comment]
No host?

**Comment by SteveL-MSFT** [comment]
Figured out why it's failing on non-Windows.  It appears that exit codes from programs have to be in the [range 0-255 and anything greater than 255 is mod 256](http://www.tldp.org/LDP/abs/html/exitcodes.html).  I verified this with a simple console c# app.  In this case, the value for ExitCodeBadCommandLineParameter mod 256 == 0.  Hmmm.

Although a breaking change, I think we should conform to [libc standards](https://opensource.apple.com/source/Libc/Libc-320/include/sysexits.h) which means this error code should be 64 "command line usage error".  We should also update all the other error codes that powershell console returns.

In general, I don't think this should be a problem as typically 0 means success and non-0 means error so although a breaking change, I suspect it have limited real world impact.

cc @PowerShell/powershell-committee

**Review by lzybkr — APPROVED** [review]
I approve the changes, but there are 2 distinct fixes in the PR, so I think there should be 2 distinct commits.

Interactive add (`git add -p`) is an easy way to do that.

**Comment by SteveL-MSFT** [comment]
@lzybkr are you referring to -WindowStyle and everything else as two commits?

**Comment by SteveL-MSFT** [comment]
@lzybkr sure, I can make that two commits in the same PR.

**Comment by SteveL-MSFT** [comment]
Split into 3 commits (Thanks @lzybkr for the tip on `git add -p`!)

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and is ok with changing the exit codes

**Review by iSazonov — APPROVED** [review]

**Comment by TravisEz13** [comment]
@iSazonov and @SteveL-MSFT  thanks as always!
